wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/LanguageServerInitializer.java

```diff
@@ -0,0 +1,228 @@
+
+  private final ExecutorService executor;
+
+  private final EventService eventService;
+  private final CheLanguageClientFactory cheLanguageClientFactory;
+  private final InitializeParamsProvider initializeParamsProvider;
+  private final ServerCapabilitiesAccumulator serverCapabilitiesAccumulator;
+  private final FindId findId;
+
+  private final Registry<String> idRegistry;
+  private final Registry<LanguageServer> languageServerRegistry;
+  private final Registry<InstanceProvider> instanceProviderRegistry;
+  private final Registry<ServerCapabilities> serverCapabilitiesRegistry;
+  private final Registry<Pair<InputStream, OutputStream>> ioStreamRegistry;
+  private final Registry<CommunicationProvider> communicationProviderRegistry;
+
+  @Inject
+  public LanguageServerInitializer(
+      ServerCapabilitiesAccumulator serverCapabilitiesAccumulator,
+      RegistryContainer registryContainer,
+      FindId findId,
+      EventService eventService,
+      CheLanguageClientFactory cheLanguageClientFactory,
+      InitializeParamsProvider initializeParamsProvider) {
+    this.executor = newCachedThreadPool(getFactory());
+
+    this.eventService = eventService;
+    this.cheLanguageClientFactory = cheLanguageClientFactory;
+    this.initializeParamsProvider = initializeParamsProvider;
+    this.serverCapabilitiesAccumulator = serverCapabilitiesAccumulator;
+    this.findId = findId;
+
+    this.idRegistry = registryContainer.idRegistry;
+    this.languageServerRegistry = registryContainer.languageServerRegistry;
+    this.instanceProviderRegistry = registryContainer.instanceProviderRegistry;
+    this.serverCapabilitiesRegistry = registryContainer.serverCapabilitiesRegistry;
+    this.ioStreamRegistry = registryContainer.ioStreamRegistry;
+    this.communicationProviderRegistry = registryContainer.communicationProviderRegistry;
+  }
+
+  private static ThreadFactory getFactory() {
+    return new ThreadFactoryBuilder()
+        .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+        .setNameFormat(LanguageServerInitializer.class.getSimpleName())
+        .setDaemon(true)
+        .build();
+  }
+
+  /**
+   * Initialize all language servers that match a specified workspace path. If some servers are
+   * already initialized does nothing. If some servers can't be initialized due to some errors at
+   * any point - skips them and proceeds with the rest.
+   *
+   * @param wsPath absolute workspace path
+   * @return accumulated server capabilities of all initialized language servers
+   * @throws CompletionException if no server initialized throws {@link LanguageServerException}
+   *     wrapped by {@link CompletionException}
+   */
+  CompletableFuture<ServerCapabilities> initialize(String wsPath) {
+    return supplyAsync(
+        () -> {
+          LOG.info("Started language servers initialization, file path '{}'", wsPath);
+
+          Set<ServerCapabilities> serverCapabilitiesSet =
```

Comment:
> As I read the code, this will start the language servers in sequence, instead of in parallel. 

Absolutely right, this will start language server in sequence, but only for a single path. Requests for several paths are processed in parallel. 

> This might be slower and may need an increased timeout when waiting for LS initialization.

I understand your point, I guess timeout should be left as is, while instead this part is parallelized.

In general that's actual only for installers/launchers approach, because in parallel containers all servers are launched at container start, as far as I understand.

Reply:
Yes, but a lot of work will only be done when "initialize" is called on the language server. Before, the language server is just waiting for a connection.